### PR TITLE
PKCE implementation

### LINKED ISSUE #1114: PKCE support for public clients
It's intended to remedy possible attacks against public clients. See [RFC7636: Proof Key for Code Exchange by OAuth Public Clients](https://tools.ietf.org/html/rfc7636). It's a small twist to the authorization code flow:

```
                                                 +-------------------+
                                                 |   Authz Server    |
       +--------+                                | +---------------+ |
       |        |--(A)- Authorization Request ---->|               | |
       |        |       + t(code_verifier), t_m  | | Authorization | |
       |        |                                | |    Endpoint   | |
       |        |<-(B)---- Authorization Code -----|               | |
       |        |                                | +---------------+ |
       | Client |                                |                   |
       |        |                                | +---------------+ |
       |        |--(C)-- Access Token Request ---->|               | |
       |        |          + code_verifier       | |    Token      | |
       |        |                                | |   Endpoint    | |
       |        |<-(D)------ Access Token ---------|               | |
       +--------+                                | +---------------+ |
                                                 +-------------------+

                     Figure 2: Abstract Protocol Flow
```

Anyone interested in supporting this? 😃

### PR DESCRIPTION
**Pull request by HEllRZA**
This pull request adds the PKCE (Proof Key for Code Exchange) implementation. ([RFC 7636](https://tools.ietf.org/html/rfc7636)), Closes #1114.

Based on the implementation of @Teeed (PR #1652)
- add the remarks of @deric (return invalid_grand when wrong code_verifier)
- fix the issue, @mfmarche found.
- Enforce PKCE flow on /token when PKCE flow was started on /auth
- made sure, the correct errors are returned
- add Tests for the error cases
- make sure, the client_secret is only omitted, when code_verifier is received on /token, and grant_type is correct.
- add "Authorization" to allowed CORS headers

I am lookin forward to your review.

### Test PKCE
You should use a public client (without secret), so PKCE is not stopped by a missing client_secret
```yaml
staticClients:
- id: example-app
  name: 'Example App'
  public: true
```
### Note
* Configuring redirectURIs does currently not work with public clients (see #1822) but public clients allow all localhost callback URIs (e.g. http://localhost:5556/dex/callback)
* When testing with oidc-client.js if you use `loadUserInfo: true`, it will not work until #1819 is merged.

Thanks to @Teeed and i hope this PR is of interest to @mfmarche, @deric, @justin-slowik

### CODE REVIEW

**Comment by Teeed** [comment]
Hi,

good job! :+1:

**Comment by HEllRZA** [comment]
@sagikazarmark could you take a look at it? That would be great!

**Comment by asoorm** [comment]
Is it necessary to support the plain code challenge method?

https://tools.ietf.org/html/rfc7636#section-7.2

>    Because of this, "plain" SHOULD NOT be used and exists only for
>    compatibility with deployed implementations where the request path is
>    already protected.  The "plain" method SHOULD NOT be used in new
>    implementations, unless they cannot support "S256" for some technical
>    reason.
> 
>    The "S256" code challenge method or other cryptographically secure
>    code challenge method extension SHOULD be used.  The "plain" code
>    challenge method relies on the operating system and transport
>    security not to disclose the request to an attacker.

https://tools.ietf.org/html/draft-ietf-oauth-v2-1-00#section-4.1.1.2

> The plain transformation is for compatibility with existing
   deployments and for constrained environments that can't use the
   "S256" transformation.

**Comment by HEllRZA** [comment]
@asoorm 
Thanks for your input! I also dislike `"plain"`.

Reading the RFC, it sounds to me that the server has to support the `code_challenge_method "plain"`. It is even the default, when no `code_challenge_method` is specified. The passages you quote discourage the use of `"plain"` in the client that starts the PKCE request.

> Client Sends the Code Challenge with the Authorization Request
>
>   The client sends the code challenge as part of the OAuth 2.0
>   Authorization Request (Section 4.1.1 of [RFC6749]) using the
>   following additional parameters:
>
>   code_challenge
>      REQUIRED.  Code challenge.
>
>   code_challenge_method
>      OPTIONAL, defaults to "plain" if not present in the request.  Code
>      verifier transformation method is "S256" or "plain".

However the client should not use it. It should use `"S256"` and never downgrade to `"plain"`. `"S256"` is required on the server.
>Clients MUST NOT downgrade to "plain" after trying the "S256" method.
>   Servers that support PKCE are required to support "S256", and servers
>   that do not support PKCE will simply ignore the unknown
>   "code_verifier".

Other Auth Providers also support both (e.g. auth0).
Attacks should be prevented by disallowing to use `"plain"` on the /token endpoint, when `"S256"` was used as `code_challenge_method` on the /auth endpoint to start the flow. **I will make sure that i have a test for that!**

File: server/handlers.go
Comment (tkleczek) [inline server/handlers.go:819]: already declared before 'if'
```suggestion
```

File: server/handlers.go
Comment (tkleczek) [inline server/handlers.go:822]: I believe according to https://tools.ietf.org/html/rfc7636#section-4.4.1, errInvalidRequest and http.StatusBadRequest should be returned here (in case of unknown code challenge method)

```
If the server supporting PKCE does not support the requested
   transformation, the authorization endpoint MUST return the
   authorization error response with "error" value set to
   "invalid_request".  The "error_description" or the response of
   "error_uri" SHOULD explain the nature of error, e.g., transform
   algorithm not supported
```

File: server/oauth2.go
Comment (tkleczek) [inline server/oauth2.go:425]: I believe the error of unsupported PKCE challenge method should be returned as a query param in the redirect to a client callback URI.
So this check should be performed after redirectURI gets validated below and the newErr func defined there should be used.

File: server/handlers.go
Comment (tkleczek) [inline server/handlers.go:766]: Is this really necessary? What's wrong with checking clientSecret even for PKCE?

File: server/oauth2.go
Comment (HEllRZA) [reply server/oauth2.go:425]: This is the passage from the RFC. I think that it is implemented according to it. What do you think?
```
RFC 7636                       OAUTH PKCE                 September 2015
4.4.1.  Error Response
   [...]
   If the server supporting PKCE does not support the requested
   transformation, the authorization endpoint MUST return the
   authorization error response with "error" value set to
   "invalid_request".  The "error_description" or the response of
   "error_uri" SHOULD explain the nature of error, e.g., transform
   algorithm not supported.
```

File: server/server_test.go
Comment (tkleczek) [inline server/server_test.go:1335]:
```suggestion
		t.Errorf("%s: unexpected error type: %s. expected *oauth2.RetrieveError", tc.name, reflect.TypeOf(err))
```

File: server/handlers.go
Comment (HEllRZA) [reply server/handlers.go:766]: If we would check the clientSecret for PKCE, the client would have to possess the client secret. PKCE is created for public clients that **cannot safely** store a clientSecret; e.g. single-page applications.

[Very good explanation of Auth Code flow with PKCE](https://auth0.com/docs/flows/authorization-code-flow-with-proof-key-for-code-exchange-pkce)

However; if someone does not want to have PKCE enabled, because he **always** wants to use the client secret, we should think about a configuration to enable/disable PKCE.

**Review by tkleczek — COMMENTED** [review]
Would be nice to also add some tests to oauth2_test.go

File: server/oauth2.go
Comment (tkleczek) [reply server/oauth2.go:425]: The thing is, oauth2 rfc https://tools.ietf.org/html/rfc6749#section-4.2.2.1 says
```
If the request fails due to a missing, invalid, or mismatching
   redirection URI, or if the client identifier is missing or invalid,
   the authorization server SHOULD inform the resource owner of the
   error and MUST NOT automatically redirect the user-agent to the
   invalid redirection URI.

   **If the resource owner denies the access request or if the request
   fails for reasons other than a missing or invalid redirection URI,
   the authorization server informs the client by adding the following
   parameters to the fragment component of the redirection URI** using the
   "application/x-www-form-urlencoded" format, per Appendix B:
```

Using authErr{"", "", errInvalidRequest, description}  (with first two params empty) would result in dex just displaying page with error instead of sending HTTP redirect back to client

File: server/handlers.go
Comment (tkleczek) [reply server/handlers.go:766]: I agree that PKCE was introduced to mitigate security concerns for public clients, but my reasoning is: 
If client is configured as public, then the check should pass as both clientSecret and client.Secret should be empty.
But if for whatever reason sb configures confidential client with PKCE, why not validate client secret as well.

File: storage/storage.go
Comment (tkleczek) [inline storage/storage.go:173]: I am wondering if the better names would be 
CodeChallenge -> Challenge
CodeChallengeMethod -> Method
so that we don't have excessive verbosity when used?: 
CodeChallenge.CodeChallengeMethod

alternatively the type could be renamed to PKCE

File: server/oauth2.go
Comment (HEllRZA) [reply server/oauth2.go:425]: Ok, now I understand. I will take a look at it.

File: storage/storage.go
Comment (HEllRZA) [reply storage/storage.go:173]: `CodeChallenge.Challenge` also sounds verbose, so
I prefer `PKCE.CodeChallenge` and `PKCE.CodeChallengeMethod`
It makes it clear, for what it is used.

File: server/handlers.go
Comment (HEllRZA) [reply server/handlers.go:766]: Dex currently disallows an empty client secret with the error:
`invalid config: Secret or SecretEnv field is required for client "example-app"`
and I don't want to mess with that decision. The secret protects e.g. the refresh_token flow and the password_grant flow.

I guess it is best to have a config parameter for the client like
`allowPKCE: true` which is false by default. This way, old configurations still use client_secret only and are not affected by the PKCE implementation.

Sounds good?

Comment (HEllRZA) [reply server/handlers.go:766]: With "public" you mean setting the client.public parameter and thus enabling https://github.com/dexidp/dex/pull/471?

### Suggestion one
```suggestion
        if client.Public && grantType == grantTypeAuthorizationCode && codeVerifier != "" {
               // let it pass
        } else // check secret
```
This will accomplish the following:

if client is confidential, the clientSecret will always be checked (
if client is public, but the client has a client secret, PKCE will still work *without* a client secret.

### Suggestion two
introduce a new boolean to client configuration, which allows PKCE without checking the client secret. This way the client does not have to be set to public.

**Comment by HEllRZA** [comment]
@tkleczek 
Thanks for the code review!
from the [RFC 7636](https://tools.ietf.org/html/rfc7636#section-4.4.1)
```
4.4.1.  Error Response
   If the server requires Proof Key for Code Exchange (PKCE) by OAuth
   public clients and the client does not send the "code_challenge" in
   the request, the authorization endpoint MUST return the authorization
   error response with the "error" value set to "invalid_request".  The
   "error_description" or the response of "error_uri" SHOULD explain the
   nature of error, e.g., code challenge required.
```

Sounds to me as if it would be a good idea to have a switch in the dex client configuration,
that sets PKCE as "required" for the code auth flow of public clients, what do you think?

File: server/handlers.go
Comment (tkleczek) [reply server/handlers.go:766]: `With "public" you mean setting the client.public parameter and thus enabling #471?`
Yes.

I am trying to understand the scenario you have in mind behind the this if-check. Are you concerned about allowing access in scenario of having a single client configuration dex-side (as confidential client) and then have 2 clients that use it - one is confidential (knows the secret and can protect it) and other is public (does not know the secret and uses PKCE)?

**Comment by tkleczek** [comment]
@HEllRZA 
 
> Sounds to me as if it would be a good idea to have a switch in the dex client configuration,
> that sets PKCE as "required" for the code auth flow of public clients, what do you think?

Yeah, for me definitely nice-to-have - but not required. I was trying to think whether to have a server-wide or per-client option. I guess server-wide default could save your time if you would like every client to be forced to use PKCE. On the other hand, some clients could not be ready for it - so having a per-client config also seems to have benefits.

File: server/handlers.go
Comment (HEllRZA) [reply server/handlers.go]: > Are you concerned about allowing access in scenario of having a single client configuration dex-side (as confidential client) and then have 2 clients that use it - one is confidential (knows the secret and can protect it) and other is public (does not know the secret and uses PKCE)?

Yes.

Comment (tkleczek) [reply server/handlers.go]: I think this use-case might violate the OIDC spec:
https://openid.net/specs/openid-connect-core-1_0.html
```
9.  Client Authentication
This section defines a set of Client Authentication methods that are used by Clients to authenticate to the Authorization Server when using the Token Endpoint. During Client Registration, the RP (Client) MAY register a Client Authentication method. If no method is registered, the default method is client_secret_basic.

These Client Authentication methods are:

client_secret_basic
Clients that have received a client_secret value from the Authorization Server authenticate with the Authorization Server in accordance with Section 2.3.1 of OAuth 2.0 [RFC6749] using the HTTP Basic authentication scheme.
[...]
none
The Client does not authenticate itself at the Token Endpoint, either because it uses only the Implicit Flow (and so does not use the Token Endpoint) or because it is a Public Client with no Client Secret or other authentication mechanism.
```

This seems to imply that there should be a single authentication method used by the client and your scenario assumes two (some client instances would use `client_secret_basic`, others `none`)

Comment (HEllRZA) [reply server/handlers.go]: Also reading through [this](https://github.com/mitreid-connect/OpenID-Connect-Java-Spring-Server/issues/1456) and [this](https://stackoverflow.com/questions/56793968/should-authorization-flow-use-both-secret-and-pkce-for-a-web-app), I agree upon that. Will change it.

Comment (HEllRZA) [reply server/handlers.go]: Four pur application, the restriction of the callback uris implied by the setting "public" does not work out.

Comment (HEllRZA) [reply server/handlers.go]: Actually, i don't quite understand, why "public" clients don't respect the configured redirectURIs.

Comment (HEllRZA) [reply server/handlers.go]: The document even states that wildcard domains should be prevented. The current "private" implementation violates this by basically implementing redirect URI of type `http[s]://localhost/*`.
But implementing it according to that spec, would mean to possibly break existing dex deployments :/

Comment (HEllRZA) [reply server/handlers.go]: I think i misread it, so i mimiized above comment. Should I implement the usage of RedirectURI in "private" in this pull request, because PKCE does not work correctly for SPAs without it, or should that be discussed in #1300?

Comment (tkleczek) [reply server/handlers.go]: when you say "private" you mean "public"? I think this should be discussed in #1300. I wonder if there are any downsides to just using confidential client with dummy password as a workaround for your case, don't see any at first glance.
Also, I guess this would be a good time to get some actual maintainer to interested in this PR

Comment (HEllRZA) [reply server/handlers.go]: > when you say "private" you mean "public"?

yes. (D'oh!)
> I think this should be discussed in #1300. I wonder if there are any downsides to just using confidential client with dummy password as a workaround for your case, don't see any at first glance.

The SPA draft states `NOT RECOMMENDED` to use a client_secret in that case. I am worried that we would carry that useless client_secret around "forever".
> Also, I guess this would be a good time to get some actual maintainer to interested in this PR

definitively!

**Review by tkleczek — APPROVED** [review]
LGTM

**Comment by tkleczek** [comment]
It would be great if this could be pulled-in, especially that it's straightforward RFC implementation of the feature that is considered as recommended for public clients protection (by OAuth 2.0 Security Best Current Practice document (https://tools.ietf.org/html/draft-ietf-oauth-security-topics-15)  Could anybody from maintainers team review? @sagikazarmark maybe?

**Comment by sagikazarmark** [comment]
The first four open tabs in my browser are Dex PRs to review. I'll get there eventually.

File: server/handlers.go
Comment (sagikazarmark) [inline server/handlers.go:788]:
```suggestion
```

File: server/handlers.go
Comment (sagikazarmark) [inline server/handlers.go:816]:
```suggestion
			s.logger.Error(err)
			s.tokenErrHelper(w, errServerError, "", http.StatusInternalServerError)
```

File: server/server.go
Comment (sagikazarmark) [inline server/server.go:290]: Is this change related to PKCE? If not, what's the reason for adding it?

Also, I don't think it's enough if you need to send credentials in a cors request. I think you also need `AllowCredentials` as well.

But if it's not related to the PR and it's there for testing reasons, it should probably be removed.

**Review by sagikazarmark — CHANGES_REQUESTED** [review]
LGTM, thanks a lot for working on it.

Found a few nits, nothing special.

File: server/server.go
Comment (HEllRZA) [reply server/server.go:290]: I added `"Authorization"` to the allowed headers that, from a browser, after a successful login with PKCE, the `/userinfo` endpoint can be called with the bearer token from the browser.
I tested it with oidc-client, which has a parameter `loadUserInfo`, which does just that successfully.

According to [this](https://www.sjoerdlangkemper.nl/2018/09/12/authorization-header-and-cors/), whitelisting `"Authorization"` in the `Access-Control-Allow-Headers` is sufficient, if we just want to send a bearer token (and do not want the browser to ask the user for a password).

But it is not essential for PKCE, i would consider it a bug. There is already an issue [1532](https://github.com/dexidp/dex/issues/1532) for it.

I will remove it the.

File: server/oauth2.go
Comment (heidemn-faro) [inline server/oauth2.go:591]: @sagikazarmark our use case is to use PKCE for a web app that doesn't run on "localhost".
Therefore, we would like that `RedirectURIs` are also considered for public clients.

The previous code to allow any localhost URI, `redirectURIOOB` or `deviceCallbackURI` for public clients is still active.
An alternative could be to no longer allow these special URLs when `RedirectURIs` is configured.
I don't have a strong opinion which option is better.

File: server/server.go
Comment (HEllRZA) [reply server/server.go]: A similar patch is handled in PR #1819

File: server/oauth2.go
Comment (tkleczek) [reply server/oauth2.go:591]: I think it might be better to add this change in a separate PR (as it is not PKCE-specific) just to get the PKCE feature merged faster

Comment (heidemn-faro) [reply server/oauth2.go:591]: @tkleczek OK, I just created a separate PR here: https://github.com/dexidp/dex/pull/1822

File: server/handlers.go
Comment (heidemn-faro) [reply server/handlers.go]: Done.

File: server/handlers.go
Comment (heidemn-faro) [reply server/handlers.go:816]: Done.

**Comment by heidemn-faro** [comment]
@bonifaido  @sagikazarmark  it seems that all requested changes have been implemented.
Anything else that needs to be addressed before this can be merged?

File: server/server.go
Comment (heidemn-faro) [reply server/server.go]: Done (reverted).

**Comment by trentis** [comment]
I could be doing something wrong but I just tried using this PR with a oidc-client-js implementation. The handleToken method would still always respond with unauthorized at the checking of client secret step. Any ideas why this would be happening?

**Comment by HEllRZA** [comment]
@trentis It was decided to still use the client_secret with PKCE. 

So there are two ways to solve this issue for you:
1. Use a client secret with oidc-client-js (that's probably not what you want to do)
2. Configure your dex-client as **public client** with **no client secret**.
```yaml
staticClients:
- id: example-app
  name: 'Example App'
  public: true
```
Note that configuring `redirectURIs` does currently not work with public clients (see  #1822) but public clients allow all localhost callback URIs (e.g. http://localhost:5556/dex/callback) (IIRC 127.0.0.1 does not work!)

**Comment by HEllRZA** [comment]
@trentis Another pitfall:
If you use ` loadUserInfo: true` in oidc-client-js, it will not work until #1819 is merged.

**Comment by trentis** [comment]
Cheers for the assistance @HEllRZA . Yeah an error might be good took me a while to realize it wasn't oidc-client-js  as they do the code challenge automatically. So it particularly could help others. I'll probably end up using a public client and merge in the other PRs to get what I need. Hopefully it gets merged soon

File: server/handlers.go
Comment (tkleczek) [inline server/handlers.go:762]: Seems inconsistent that we want to specifically call out "missing credentials" in PKCE case and not in general case. 
Also I'm wondering whether not specifying client authentication error details but just general "Invalid client credentials." is for security reasons. E.g. the error you introduced would allow to determine whether the client having a specific client_id is registered on the server.

Comment (HEllRZA) [reply server/handlers.go:762]: I understand. Should we instead just log it?

Comment (tkleczek) [reply server/handlers.go:762]: I would say it's a  good idea, many such entries could indicate that some attack is in place. We could also log when client_id is invalid. Not sure if we would like to specifically call-out "missing credentials" case, but I'm not against it. And let's do it regardless of PKCE.

Comment (HEllRZA) [reply server/handlers.go:762]: Ok, Now, there is a log output instead.

Comment (tkleczek) [reply server/handlers.go:762]: I was thinking more in lines of:

```
           if client.Secret != clientSecret {
	        if clientSecret != "" {
	            s.logger.Info("missing client secret")
	        } else {
	            s.logger.Info("invalid client secret")
	        }
	        
		s.tokenErrHelper(w, errInvalidClient, "Invalid client credentials.", http.StatusUnauthorized)					 
                return
```

Comment (HEllRZA) [reply server/handlers.go:762]: Written text has so much potential for misunderstanding ;o)

**Comment by sagikazarmark** [comment]
@tkleczek @HEllRZA is this PR ready for another look?

**Comment by tkleczek** [comment]
@sagikazarmark  Yes, I believe so.

**Comment by HEllRZA** [comment]
@tkleczek @sagikazarmark 
Yes, it is ripe.

File: server/handlers.go
Comment (johanbrandhorst) [inline server/handlers.go:764]: Should this case be OK for PKCE? The whole idea is that the client doesn't have to provide the client secret. The role of the IdP in this case is just to verify Code verifier and Code challenge, which happens in `handleAuthCode`.

Ref: https://auth0.com/docs/flows/authorization-code-flow-with-proof-key-for-code-exchange-pkce

Comment (johanbrandhorst) [reply server/handlers.go:764]: I tried rewriting this locally to move the clientSecret validation into `handleAuthCode`, `handleRefreshToken` and `handlePasswordGrant` respectively. I think we have to remove it from this function for PKCE to work.

Comment (johanbrandhorst) [reply server/handlers.go:764]: Here's a rought sketch of the changes I made, though I'd suggest extracting the client secret validation into its own function instead:
```diff
diff --git a/server/handlers.go b/server/handlers.go
index 342849ee..1141dffc 100644
--- a/server/handlers.go
+++ b/server/handlers.go
@@ -765,24 +765,15 @@ func (s *Server) handleToken(w http.ResponseWriter, r *http.Request) {
 		}
 		return
 	}
-	if client.Secret != clientSecret {
-		if clientSecret == "" {
-			s.logger.Infof("missing client_secret on token request for client: %s", client.ID)
-		} else {
-			s.logger.Infof("invalid client_secret on token request for client: %s", client.ID)
-		}
-		s.tokenErrHelper(w, errInvalidClient, "Invalid client credentials.", http.StatusUnauthorized)
-		return
-	}
 
 	grantType := r.PostFormValue("grant_type")
 	switch grantType {
 	case grantTypeAuthorizationCode:
-		s.handleAuthCode(w, r, client)
+		s.handleAuthCode(w, r, client, clientSecret)
 	case grantTypeRefreshToken:
-		s.handleRefreshToken(w, r, client)
+		s.handleRefreshToken(w, r, client, clientSecret)
 	case grantTypePassword:
-		s.handlePasswordGrant(w, r, client)
+		s.handlePasswordGrant(w, r, client, clientSecret)
 	default:
 		s.tokenErrHelper(w, errInvalidGrant, "", http.StatusBadRequest)
 	}
@@ -801,7 +792,7 @@ func (s *Server) calculateCodeChallenge(codeVerifier, codeChallengeMethod string
 }
 
 // handle an access token request https://tools.ietf.org/html/rfc6749#section-4.1.3
-func (s *Server) handleAuthCode(w http.ResponseWriter, r *http.Request, client storage.Client) {
+func (s *Server) handleAuthCode(w http.ResponseWriter, r *http.Request, client storage.Client, clientSecret string) {
 	code := r.PostFormValue("code")
 	redirectURI := r.PostFormValue("redirect_uri")
 
@@ -839,6 +830,16 @@ func (s *Server) handleAuthCode(w http.ResponseWriter, r *http.Request, client s
 		// Received PKCE request on /auth, but no code_verifier on /token
 		s.tokenErrHelper(w, errInvalidGrant, "Expecting parameter code_verifier in PKCE flow.", http.StatusBadRequest)
 		return
+	} else {
+		if client.Secret != clientSecret {
+			if clientSecret == "" {
+				s.logger.Infof("missing client_secret on token request for client: %s", client.ID)
+			} else {
+				s.logger.Infof("invalid client_secret on token request for client: %s", client.ID)
+			}
+			s.tokenErrHelper(w, errInvalidClient, "Invalid client credentials.", http.StatusUnauthorized)
+			return
+		}
 	}
 
 	if authCode.RedirectURI != redirectURI {
@@ -1000,7 +1001,17 @@ func (s *Server) exchangeAuthCode(w http.ResponseWriter, authCode storage.AuthCo
 }
 
 // handle a refresh token request https://tools.ietf.org/html/rfc6749#section-6
-func (s *Server) handleRefreshToken(w http.ResponseWriter, r *http.Request, client storage.Client) {
+func (s *Server) handleRefreshToken(w http.ResponseWriter, r *http.Request, client storage.Client, clientSecret string) {
+	if client.Secret != clientSecret {
+		if clientSecret == "" {
+			s.logger.Infof("missing client_secret on token request for client: %s", client.ID)
+		} else {
+			s.logger.Infof("invalid client_secret on token request for client: %s", client.ID)
+		}
+		s.tokenErrHelper(w, errInvalidClient, "Invalid client credentials.", http.StatusUnauthorized)
+		return
+	}
+
 	code := r.PostFormValue("refresh_token")
 	scope := r.PostFormValue("scope")
 	if code == "" {
@@ -1227,7 +1238,17 @@ func (s *Server) handleUserInfo(w http.ResponseWriter, r *http.Request) {
 	w.Write(claims)
 }
 
-func (s *Server) handlePasswordGrant(w http.ResponseWriter, r *http.Request, client storage.Client) {
+func (s *Server) handlePasswordGrant(w http.ResponseWriter, r *http.Request, client storage.Client, clientSecret string) {
+	if client.Secret != clientSecret {
+		if clientSecret == "" {
+			s.logger.Infof("missing client_secret on token request for client: %s", client.ID)
+		} else {
+			s.logger.Infof("invalid client_secret on token request for client: %s", client.ID)
+		}
+		s.tokenErrHelper(w, errInvalidClient, "Invalid client credentials.", http.StatusUnauthorized)
+		return
+	}
+
 	// Parse the fields
 	if err := r.ParseForm(); err != nil {
 		s.tokenErrHelper(w, errInvalidRequest, "Couldn't parse data", http.StatusBadRequest)
```

Comment (HEllRZA) [reply server/handlers.go:764]: @johanbrandhorst

I initially implemented it in a similar way by skipping the client_secret validation, if the request was PKCE.

We decided that, for that case of using PKCE, the client should be configured as **public** and have no client_secret. And we should not omit the client_secret if the client is **confidential**.

Quote from @tkleczek:
> I agree that PKCE was introduced to mitigate security concerns for public clients, but my reasoning is:
> If client is configured as public, then the check should pass as both clientSecret and client.Secret should be empty.
> But if for whatever reason sb configures confidential client with PKCE, why not validate client secret as well.

Comment (johanbrandhorst) [reply server/handlers.go:764]: That makes sense, thanks!

**Review by sagikazarmark — APPROVED** [review]
LGTM, thanks @HEllRZA and @tkleczek for your work! ❤️

**Review by bonifaido — APPROVED** [review]
LGTM, thank you very much!

**Pull request merged**
